src/shared/source-handling.ts

```diff
@@ -0,0 +1,93 @@
+import { createHmac, timingSafeEqual } from "crypto";
+import { handler as collectMetricsHandler } from "../handler";
+import { RawEvent, TriggerSource, headerSourceMap } from "../interfaces";
+import { LogErrors } from "./log-messages";
+
+interface EventHeaders {
+  [name: string]: string | undefined;
+}
+
+const GITHUB_SIGNATURE_HEADER = "x-hub-signature-256";
+
+export function identifySourceAndEvent(
+  headers: EventHeaders
+): Pick<RawEvent, "source" | "sourceEventSignature"> {
+  for (const [headerName, source] of Object.entries(headerSourceMap)) {
+    const sourceEventSignature = headers[headerName];
+
+    if (sourceEventSignature) {
+      return { source, sourceEventSignature };
+    }
+  }
+
+  return {
+    source: TriggerSource.Unknown,
+    sourceEventSignature: "unknown",
+  };
+}
+
+export function validateRequest(
+  headers: EventHeaders,
+  body: string,
+  webhookSecret: string
+): { statusCode: number; message: string } | undefined {
+  // Github signature header must be present
+  if (!headers || !headers[GITHUB_SIGNATURE_HEADER]) {
+    return { statusCode: 400, message: LogErrors.badRequestError };
+  }
+  const githubSignature = headers[GITHUB_SIGNATURE_HEADER];
+
+  // Verify this request by checking its signature
+  if (!verifyGithubSignature(githubSignature, body, webhookSecret)) {
+    return { statusCode: 401, message: LogErrors.forbiddenRequestError };
+  }
+}
```

Comment:
I don't love this function API ("It returns an error if something is wrong, or undefined if not"), because the "main" return value of a function is usually the happy path. The implicit return of undefined does not signal an "all good" message to me when reading the code.

Looking at the code where this function is called, I believe you could `throw` an `HTMLError` directly instead of returning it?

Reply:
Yeah I know what you mean.
I only ported the function but also don't like that it only returns != undefined in an error case.
Your suggestion should work, will give it a try.